atomic_reactor/plugins/pre_fetch_sources.py

```diff
@@ -348,3 +365,126 @@ def get_signing_intent(self):
 
         signing_intent = odcs_config.get_signing_intent_by_name(self.signing_intent)
         return signing_intent
+
+    def _get_exclude_list(self, request_session, exclude_list_url):
+        response = request_session.get(exclude_list_url)
+        response.raise_for_status()
+        exclude_list_yaml = yaml.safe_load(response.text)
+        # prepend os.sep so it starts with exact match, glob.glob doesn't need it
+        return [os.path.join(os.sep, k, item) for k, v in exclude_list_yaml.items() for item in v]
+
+    def _create_full_remote_sources_map(self, request_session, remote_sources_map,
+                                        remote_sources_dir):
+        full_remote_sources_map = {}
+
+        for remote_source, remote_source_json in remote_sources_map.items():
+            remote_source_archive = os.path.join(remote_sources_dir, remote_source)
+
+            response = request_session.get(remote_source_json)
+            response.raise_for_status()
+            response_json = response.json()
+            full_remote_sources_map[remote_source_archive] = response_json
+        return full_remote_sources_map
+
+    def _check_if_package_excluded(self, remote_json, exclude_list, remote_archive):
+        delete_app = False
+        # check if any package in cachito json matches excluded entry
+        for package in remote_json['packages']:
+            if delete_app:
+                break
+
+            for exclude_path in exclude_list:
+                if package.get('name').endswith(exclude_path):
+                    self.log.debug('Package excluded: "%s" from "%s"', package.get('name'),
+                                   remote_archive)
+                    delete_app = True
+                    break
+        return delete_app
+
+    def _delete_app_directory(self, remote_source_dir, unpack_dir, remote_archive):
+        vendor_dir = os.path.join(remote_source_dir, unpack_dir, 'app', 'vendor')
+
+        if os.path.exists(vendor_dir):
+            shutil.move(vendor_dir, remote_source_dir)
+            self.log.debug('Removing app from "%s"', remote_archive)
+            shutil.rmtree(os.path.join(unpack_dir, 'app'))
+            shutil.move(os.path.join(remote_source_dir, 'vendor'), vendor_dir)
```

Comment:
Again, I might be missing something, but how can you move the vendor dir to `/somewhere/app/vendor` if `/somewhere/app/` was just deleted?

Reply:
vendor_dir   :    /somewhere/unpacked_dir/app/vendor
first we move 'vendor' to /somewhere/
then we delete : /somewhere/unpacked_dir/app/vendor
then we move vendor back   to /somewhere/unpacked_dir/app/vendor  from /somewhere/vendor

Reply:
If I undestand it correctly, the entire process is similar to this (`tree` command just for visualization):
```shell
$ tree 'remote_sources'
remote_sources
└── foo_unpacked
    └── app
        └── vendor
$ mv 'remote_sources/foo_unpacked/app/vendor' 'remote_sources'
$ tree 'remote_sources'
remote_sources
├── foo_unpacked
│   └── app
└── vendor
$ rm -rf 'remote_sources/foo_unpacked/app'
$ tree 'remote_sources'
remote_sources
├── foo_unpacked
└── vendor
$ mv 'remote_sources/vendor' 'remote_sources/foo_unpacked/app/vendor'
mv: cannot move 'remote_sources/vendor' to 'remote_sources/foo_unpacked/app/vendor': No such file or directory
```
If that is correct, then how come the last `shutil.move` command does not fail? From the little test that I tried, `shutil.move` does not seem to create missing directories along the way, just like `mv` doesn't

Reply:
yes that is correct but shutil.move works:
here is example:

>>> shutil.move('/tmp/tmpo44pO_/remote_sources/unpacked/app/vendor', '/tmp/tmpo44pO_/remote_sources/')
>>> shutil.rmtree('/tmp/tmpo44pO_/remote_sources/unpacked/app')
>>> shutil.move('/tmp/tmpo44pO_/remote_sources/vendor', '/tmp/tmpo44pO_/remote_sources/unpacked/app/vendor')

Reply:
shutil.move(src, dst)
Recursively move a file or directory (src) to another location (dst).

Reply:
Oh ok, so there seems to be a difference between how `shutil.move` handles directories vs. simple files:
```python
>>> shutil.move('/tmp/foo.txt', '/tmp/nonexistent/foo.txt')
FileNotFoundError: [Errno 2] No such file or directory: '/tmp/nonexistent/foo.txt'
>>> shutil.move('/tmp/bar/', '/tmp/nonexistent/bar')
'/tmp/nonexistent/bar'
```
Could you add a comment that `shutil.move` will create missing parent directories in `dst` when `src` is a directory?

Reply:
sure